Add Crownstone integration

RicArch97 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

New integration for [Crownstone](https://crownstone.rocks)!

Crownstone creates smart crown stones to build into a power outlet, and plugs that plug into regular outlets/power strips.
A standout feature is the ability to detect a user's presence on room level by using BLE.

The integration exists of 3 core external libraries.
- [crownstone-uart](https://github.com/crownstone/crownstone-lib-python-uart) (Adds the ability to use the [Crownstone USB](https://shop.crownstone.rocks/products/crownstone-usb-dongle) with Home Assistant. This dongle hooks directly into the Bluetooth mesh.)
- [crownstone-cloud](https://github.com/crownstone/crownstone-lib-python-cloud) (Pulls all data from our Cloud Service and adds switching/dimming commands)
- [crownstone-sse](https://github.com/crownstone/crownstone-lib-python-sse) (Awesome server-send-events client to listen for changes in presence, and more to come!)

All of these libraries are available on PyPi.

Integration is configurable via the UI.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/13791
- Link to brands pull request: https://github.com/home-assistant/brands/pull/1712

The full Crownstone integration (so far) is now available on HACS for testing.
See [Crownstone HACS repository](https://github.com/crownstone/crownstone-home-assistant).

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

RicArch97 committed:
Initial commit for Crownstone integration

RicArch97 committed:
Adjust all code to pass lint tests. Fixed state updates of Crownstones. Fixed brightness adjustment. Fixed bug where presence was not updated correctly.

RicArch97 committed:
Bumped Crownstone dependencies versions

RicArch97 committed:
Bumped crownstone-uart version

RicArch97 committed:
Updated crownstone-cloud dependency to 1.2.1. Updated .coveragerc.

homeassistant commented:
Hi @RicArch97, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#37005).

Once you do that we will be able to review and accept this pull request.

Thanks!

RicArch97 committed:
Fixed pylint errors

RicArch97 committed:
Fixed small bug in light supported_features returning None

RicArch97 committed:
Updated config_flow to skip the sphere step when only 1 sphere is available, bumped crownstone-sse. Other minor cleanups/fixes.

frenck commented on .coveragerc:
```diff
@@ -145,6 +145,7 @@ omit =
     homeassistant/components/cppm_tracker/device_tracker.py
     homeassistant/components/cpuspeed/sensor.py
     homeassistant/components/crimereports/sensor.py
+    homeassistant/components/crownstone/*
```
Please do not exclude `config_flow.py` from the test coverage.

We require a configuration flow to have 100% test coverage. Currently, there are no tests in this PR, so that needs to be added.

RicArch97 replied:
Thanks for the review, i'll update the PR with the config flow test

frenck commented on homeassistant/components/crownstone/__init__.py:
```diff
@@ -0,0 +1,34 @@
+    crownstone_hub = CrownstoneHub(hass, entry)
+
+    if not await crownstone_hub.async_setup():
+        return False
```
Should this abort or try again later?
In case it should try again later, ConfigEntryNotReady should be raised.

RicArch97 replied:
Within the `async_setup` method, i already raise ConfigEntryNotReady upon an unknown error, however i don't know if this is the correct usage
```python
except CrownstoneUnknownError:
        _LOGGER.error("Unknown error during login")
        raise ConfigEntryNotReady
```
My idea was, if it's an unknown error it could be caused by the server and should be retried later. If `async_setup` returns `False` something must have gone wrong during authentication:
```python
except CrownstoneAuthenticationError as auth_err:
        _LOGGER.error(
           "Auth error during login with type: %s and message: %s",
           auth_err.type,
           auth_err.message,
        )
        return False
```
Which is quite unlikely, but in that case it should abort and the user should re-authenticate.

frenck commented on homeassistant/components/crownstone/hub.py:
```diff
@@ -0,0 +1,186 @@
+        self.cloud = CrownstoneCloud(
+            email=customer_email,
+            password=customer_password,
+            loop=self.hass.loop,
```
Do not pass in the loop

frenck commented on homeassistant/components/crownstone/translations/nl.json:
```diff
@@ -0,0 +1,32 @@
+{
```
Do not provide additional translations. Translations are managed via Lokalise.

frenck reviewed (commented):
Did a quick scan of the PR to point out some early things.
The main missing part is the tests for the configuration flow, which we do require.

Furthermore, please limit the PR to a single platform (e.g., light or sensor). Additional platforms can be added later.

RicArch97 committed:
Added config_flow test. Removed sensor platform for initial Crownstone PR.

RicArch97 commented:
Changes were made after review. To sum up the changes:
- Test for configuration flow added with 100% code coverage
- Removed the sensor platform for this first PR (only light platform remains)
- Added ability change event (belongs to light, that way users can turn dimming on/off without requiring a restart)
- Removed the hass loop parameter in CrownstoneCloud
- Removed additional translations

frenck commented on homeassistant/components/crownstone/light.py:
```diff
@@ -0,0 +1,173 @@
+    @property
+    def should_dim(self) -> bool:
+        """Return if this crownstone is able to dim."""
+        return self.crownstone.abilities.get(DIMMING_ABILITY).is_enabled
```
The change in abilities affects the `supported_features` dynamically.
This is not allowed.

RicArch97 replied:
What is the reason this isn't allowed?
Changing abilities on the fly is an important part of Crownstones.
Not supporting brightness by default will make dimming impossible, and leaving it on 1 by default will cause misunderstandings of whether it is enabled or not, e.g. people trying to set a brightness but the Crownstone just turns on at 100% brightness.
I have the Crownstone integration in use on my Raspberry Pi for a couple weeks as custom component, and i have not experienced any problems with dynamically changing `supported_features`.

What is your recommendation of approaching this problem differently?

frenck replied:
Determine the abilities upfront and keep it static for the lifetime.

RicArch97 committed:
Reload config entry when an ability is updated, instead of dynamically changing supported_features. Use assumed_state because there's no polling for state update confirmation.

RicArch97 commented:
New commit includes:
- Not dynamically changing supported features. Since the light entities DO NOT implement an update method, all the data that gets loaded into the memory in the entry setup, is static, and will never change. Whenever an update event is received that an ability has been changed through the app, the config entry is **reloaded**, causing the entities to be removed and re-added.
- Added `assumed_state` property to light, because there is no polling for state updates. This is because the state of Crownstones isn't always known/correct in the cloud.

timmo001 commented:
@RicArch97 Black has recently updated. Do a `pip install --upgrade black` and reformat your code, then the checks should fully pass

RicArch97 committed:
Updated black & reformatted code.

RicArch97 commented:
Black seems to be fine, but now checks are failing (that passed before) in files that either aren't mine, or i haven't changed anything to.

epenet reviewed (changes requested):
May I suggest that you add some extra common keys to the config_flow strings, as discussed on #40578?
already_configured, password, username, cannot_connect are all available.

RicArch97 committed:
Changed config flow strings to common keys.

RicArch97 commented:
Rebased on the lastest dev branch, after some failed checks on my previous commit.

RicArch97 commented:
Not stale. Crownstone integration has been further developed and updated [here](https://github.com/crownstone/crownstone-home-assistant) (HACS custom)

balloob commented on CODEOWNERS:
```diff
@@ -86,6 +86,7 @@ homeassistant/components/coronavirus/* @home_assistant/core
 homeassistant/components/counter/* @fabaff
 homeassistant/components/cover/* @home-assistant/core
 homeassistant/components/cpuspeed/* @fabaff
+homeassistant/components/crownstone/* @Crownstone @RicArch97
```
Are you affiliated with Crownstone?

Are they aware they are set as a code owner?

RicArch97 replied:
I started developing this integration as part of an internship for Crownstone. After it ended I decided to keep developing and maintaining the integration, but not as official member. This means Crownstone is the official owner of it, and they will take over maintenance in case I am unavailable.

mrquincle replied:
We are fine with it. ~CEO Crownstone.

balloob commented on homeassistant/components/crownstone/helpers.py:
```diff
@@ -0,0 +1,31 @@
+    def __init__(self) -> None:
+        """Init with new event loop and instance."""
+        self.loop = asyncio.new_event_loop()
+        self.uart_instance = CrownstoneUart(self.loop)
```
If this is an async package, why are you running it in a new event loop?

RicArch97 replied:
The Uart lib is created by the Crownstone core members, with the idea to have it block until a USB is connected, so it works in a recursive way instead of creating tasks in the event loop to allow for other coroutines to be executed. If no USB is connected, it just keeps on scanning for USB dongles until it finds one. This way it would mean the USB could be connected and removed at any time, and the `async_turn_on` uses the the USB whenever it's available, and falls back to the cloud if it is not.

We've decided to have it run in a separate thread to stop it from blocking the event loop.

I can request changes, if you deem it necessary.

frenck replied:
This will actually try to connect to every serial port it can find, to see if do a successful Crowdstone handshake.

<https://github.com/crownstone/crownstone-lib-python-uart/blob/master/crownstone_uart/core/uart/UartManager.py#L90>

This is very dangerous and can interfere with other equipment in use. Since this type of communication is not uncommon, I would say this is not a good method to have in Home Assistant.

balloob replied:
Holy moly. Yeah that is 100% not allowed in HA. USB port need to be passed in.

RicArch97 replied:
It is possible for the lib to take a port parameter and skip the handshake section. How do we ensure to provide the correct port though? Let's say I have the Crownstone dongle at /dev/ttyUSB0, but then i add an other USB dongle of some kind and restart HA, and the Crownstone dongle is now at /dev/ttyUSB1

frenck replied:
By not using those ports. We advise our users to use the linked `by-id` paths in general to avoid those issues.

hillstub replied:
Another approach is to use a udev rule.  In my case, I have the following line in a file in the folder `/etc/udev/rules.d/`:

```
SUBSYSTEM=="tty", ATTRS{idVendor}=="10c4", ATTRS{idProduct}=="ea60", SYMLINK+="crownstone"
```

This makes the usb dongle available at port `/dev/crownstone`

frenck replied:
@hillstub That is a modification done on the system administration level. From a Python application perspective, that is not an option or requirement we can take into account. Let's keep this PR review on the topic of reviewing code at hand. Thanks 👍

balloob commented on homeassistant/components/crownstone/hub.py:
```diff
@@ -0,0 +1,157 @@
+        self.sse.start()
+
+        # subscribe to Crownstone ability updates
+        self.sse.add_event_listener(EVENT_ABILITY_CHANGE_DIMMING, self.update_ability)
```
Is SSE calling event listeners from the event loop or a thread?

RicArch97 replied:
They are called from a thread.

balloob replied:
Then don't decorate with `@callback` and don't call any `async` function.

RicArch97 replied:
is it safe to initialize the SSE thread with the mainthread loop like so
```python
self.sse = CrownstoneSSE(
      customer_email, customer_password, asyncio.get_running_loop()
)
```
And then in the eventbus, provide coroutines to the loop by calling
```python
asyncio.run_coroutine_threadsafe(coroutine, loop)
```

balloob replied:
If it is a thread, why do you have to pass in the loop?

RicArch97 replied:
The SSE lib has it's own eventbus. Which means any callbacks or coroutines added to it are executed within that thread (and the event loop running there). Let's say i want to run
```python
await hass.async_create_task(coroutine)
```
in a listener in that thread. Do i need to pass it to the mainthread loop specifically or am I not understanding it correctly?

balloob commented on homeassistant/components/crownstone/hub.py:
```diff
@@ -0,0 +1,157 @@
+    def update_ability(self, ability_event: AbilityChangeEvent) -> None:
+        """Update the ability information."""
+        # make sure the sphere matches current.
+        update_sphere = self.cloud.spheres.find_by_id(ability_event.sphere_id)
```
Is this doing I/O ?

RicArch97 replied:
No I/O, it looks for a specific ID in dictionary data

balloob commented on homeassistant/components/crownstone/hub.py:
```diff
@@ -0,0 +1,157 @@
+        """Update the ability information."""
+        # make sure the sphere matches current.
+        update_sphere = self.cloud.spheres.find_by_id(ability_event.sphere_id)
+        if update_sphere.cloud_id == self.sphere.cloud_id:
```
use a guard clause to simplify the code.

```python
if update_sphere.cloud_id != self.sphere.cloud_id:
    return
```

balloob commented on homeassistant/components/crownstone/hub.py:
```diff
@@ -0,0 +1,157 @@
+            update_crownstone = self.sphere.crownstones.find_by_uid(
+                ability_event.unique_id
+            )
+            if update_crownstone is not None:
```
```python
            if update_crownstone is None:
                return
```

balloob commented on homeassistant/components/crownstone/hub.py:
```diff
@@ -0,0 +1,157 @@
+                    ability_event.ability_type
+                ].is_enabled = ability_event.ability_enabled
+                # reload the config entry to process the change in supported features
+                self.hass.async_create_task(
```
async_create_task can only be called from a callback inside the Home Assistant event loop.

balloob commented on homeassistant/components/crownstone/light.py:
```diff
@@ -0,0 +1,166 @@
+    entities = []
+    for crownstone in crownstone_hub.sphere.crownstones:
+        # some don't support light features
+        if crownstone.type not in CROWNSTONE_EXCLUDE:
```
It's more future proof to use an INCLUDE list.

balloob commented on homeassistant/components/crownstone/light.py:
```diff
@@ -0,0 +1,166 @@
+        """Turn off this device via dongle or cloud."""
+        if self.uart.is_ready():
+            # switch using crownstone usb dongle
+            self.uart.switch_crownstone(self.unique_id, on=False)
```
This is doing I/O inside the event loop

balloob commented on homeassistant/components/crownstone/light.py:
```diff
@@ -0,0 +1,166 @@
+                    "Dimming is not enabled for this crownstone. Go to the crownstone app to enable it"
+                )
+        elif self.uart.is_ready():
+            self.uart.switch_crownstone(self.unique_id, on=True)
```
I/O in event loop

balloob commented on homeassistant/components/crownstone/strings.json:
```diff
@@ -0,0 +1,30 @@
+        "data": {
+          "email": "[%key:common::config_flow::data::email%]",
+          "password": "[%key:common::config_flow::data::password%]"
+        },
+        "description": "Enter your email and password for Crownstone.",
+        "title": "Crownstone"
```
Title is filled in automatically based on name in manifest. This description is not needed.
```suggestion
        }
```

balloob commented on homeassistant/components/crownstone/strings.json:
```diff
@@ -0,0 +1,30 @@
+  "title": "Crownstone",
+  "config": {
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::single_instance_allowed%]"
```
Why is only a single instance allowed? I didn't see any code that required this.

RicArch97 replied:
This has to do with the Crownstone architecture. A config entry represents a "Sphere". A Sphere could be seen a house, or office. Since a single HA instance represents a house as well, it would not be logical to allow for multiple Spheres per HA instance.

balloob replied:
I think it's totally reasonable to be able to control my office space before I arrive at work or vice versa. We should not make that decision as developers. We should only limit to a single instance if a second instance is not necessary, like with Cast, which scans the network and find all.

RicArch97 replied:
There are some difficulties when trying to allow for multiple spheres in a HA instance.

Crownstone communicates through a BLE mesh network and uses the smartphone of the user as gateway to provide the data to a cloud server;  This means, if you want to switch a Crownstone that is in a different sphere, your smartphone would have to be physically there to be able to provide the data from to cloud to the BLE mesh, or in the other direction, send sensor data from the BLE mesh to the cloud server. A hub device is currently in development to help resolve that issue, but it is not ready yet.

balloob replied:
Sure, so those entities can become unavailable, just like what happens if the phone leaves the "sphere" now. That is unrelated to the number of spheres that can be configured.

balloob commented on homeassistant/components/crownstone/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+        )
+
+    async def async_step_sphere(self, user_input=None):
+        """Handle the step for selecting a sphere."""
```
We should just add all spheres. Users can always disable devices later.

balloob commented on homeassistant/components/crownstone/config_flow.py:
```diff
@@ -0,0 +1,113 @@
+            title=self.unique_id,
+            data={
+                CONF_ID: self.unique_id,
+                CONF_EMAIL: self.login_info[CONF_EMAIL],
+                CONF_PASSWORD: self.login_info[CONF_PASSWORD],
```
Since you're involved with Crownstone, can you ask them to update their authentication to use something like OAuth2? It's highly prefered over storing email/password. (not for this PR)

RicArch97 commented:
This pull request is from some time ago, and is not completely in sync anymore with what is currently available on our HACS repository [here](https://github.com/crownstone/crownstone-home-assistant). Would you recommend merging it first and then fixing according to review? (still limiting to 1 platform of course)

balloob commented:
Let's get this PR fixed and merged, and we can go from there.

balloob commented:
This PR seems to have gone stale. Closing it.

mrquincle commented:
Hi @balloob. Do you want us to open a new PR? 

We've been working on this steadily, following up your comments. It requires us to update our python code over the ecosystem in tandem though. My latest comment is from two days ago. :-)

balloob commented:
I did see your latest comment which actually made me realize this PR was stale. The developer hasn't been active on this in 1.5 months.

Feel free to open a new PR once all comments have been addressed 👍

RicArch97 commented:
The changes are still a WIP indeed, but the process is going a bit slow right now, as i have decided to refactor some code in the Python libraries. This way we start off with a very stable PR we can build upon.
